### PR TITLE
Initial Lucca.Logs commit

### PR DESCRIPTION
**Pull request by rducom**
Initial version
- use Microsoft.Extensions.Logging for native aspnet integration
- use Microsoft.Extensions.Options for native setting management
- implements default Lucca log format
- Retreive Exceptional Guids and pass to NLog

Missing for now : 
- Let users inject custom NLog & Exceptional settings with stadard IOption<T>
- e2e test with fully mocked IHttpContextAccessor
- LogLevel management for each inner loggers
- Complete usage doc (once general behavior / api surface is validated)

Questions pending :
- Do we really need to support net461 ?

### CODE REVIEW

File: Lucca.Logs/LuccaLogsOptions.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+        /// <summary>
+        /// Default application name
+        /// </summary>
+        public string AplicationName { get; set; }
+
+        /// <summary>
+        /// Exceptional Connexion String
+        /// </summary>
+        public string ConnexionString { get; set; }
```
Comment (GMouron) [inline Lucca.Logs/LuccaLogsOptions.cs]: C'est connection pas connexion

File: Lucca.Logs/LuccaLogger.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		}
+
+		/// <summary>
+		/// Is logging enabled for this logger at this <paramref name="logLevel"/>?
+		/// </summary>
+		private bool IsNlogEnabled(NLog.LogLevel logLevel)
+			=> _logger.IsEnabled(logLevel);
+    }
+
+    internal static class NlogHelper
```
Comment (GMouron) [inline Lucca.Logs/LuccaLogger.cs]: Ici on a épelé Nlog alors que les classes s'appellent NLog.
Tu as fait ça par rapport aux règles Sonar ?

File: Lucca.Logs/LogConfigurationException.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace Lucca.Logs
+{
+    public class LogConfigurationException : Exception
```
Comment (LuccaIntegration) [inline Lucca.Logs/LogConfigurationException.cs]: MAJOR Update this implementation of 'ISerializable' to conform to the recommended serialization pattern. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3925)

File: Lucca.Logs/LuccaLoggerExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            RegisterProvider(services);
+            return services;
+        }
+
+        /// <summary>
+        /// Apply NLog configuration from XML config.
+        /// </summary>
+        /// <param name="fileName">absolute path  NLog configuration file.</param>
+        /// <returns>LoggingConfiguration for chaining</returns>
+        private static LoggingConfiguration ConfigureNLog(string fileName)
```
Comment (LuccaIntegration) [inline Lucca.Logs/LuccaLoggerExtensions.cs]: MAJOR Remove the unused private method 'ConfigureNLog'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: Lucca.Logs/LuccaDataWrapper.cs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        public static class LogFilter
+        {
+            public static Func<Exception, bool> ExceptionFilter;
+        }
+
+        public static class LogExtractor
+        {
+            public static Func<Exception, IEnumerable<KeyValuePair<string, string>>> CustomKeys;
```
Comment (LuccaIntegration) [inline Lucca.Logs/LuccaDataWrapper.cs]: CRITICAL Change the visibility of 'CustomKeys' or make it 'const' or 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2223)

File: Lucca.Logs/LuccaDataWrapper.cs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        [Flags]
+        internal enum UriPart
```
Comment (LuccaIntegration) [inline Lucca.Logs/LuccaDataWrapper.cs]: MINOR Rename this enumeration to match the regular expression: '^([A-Z]{1,3}[a-z0-9]+)*([A-Z]{2})?s$'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2342)

File: Lucca.Logs/LuccaDataWrapper.cs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        {
+            None = 0,
+            Scheme = 1 << 0,
```
Comment (LuccaIntegration) [inline Lucca.Logs/LuccaDataWrapper.cs]: MINOR Remove this useless shift by 0. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2183)

File: Lucca.Logs/LuccaDataWrapper.cs [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+            if ((uriPart & UriPart.Query) == UriPart.Query)
+            {
+                urlBuilder.Append(httpRequest.QueryString.Value);
+            }
+            return urlBuilder.ToString();
+        }
+
+        public static class LogFilter
+        {
+            public static Func<Exception, bool> ExceptionFilter;
```
Comment (LuccaIntegration) [inline Lucca.Logs/LuccaDataWrapper.cs]: CRITICAL Change the visibility of 'ExceptionFilter' or make it 'const' or 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2223)

File: Lucca.Logs/LuccaLogsProvider.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using NLog;
+using StackExchange.Exceptional;
+using ILogger = Microsoft.Extensions.Logging.ILogger;
+
+namespace Lucca.Logs
+{
+	public class LuccaLogsProvider : ILoggerProvider
```
Comment (LuccaIntegration) [inline Lucca.Logs/LuccaLogsProvider.cs]: MAJOR Fix this implementation of 'IDisposable' to conform to the dispose pattern. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3881)

File: Lucca.Logs/LuccaLogsProvider.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+			_httpContextAccessor = httpContextAccessor;
+
+			LogManager.Configuration = _options.Nlog;
+            Exceptional.Configure(_options.Exceptional);
+        }
+
+		public ILogger CreateLogger(string categoryName)
+            => new LuccaLogger(categoryName, _httpContextAccessor, LogManager.GetLogger(categoryName), _options, string.Empty);
+
+        public void Dispose()
```
Comment (LuccaIntegration) [inline Lucca.Logs/LuccaLogsProvider.cs]: CRITICAL Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1186)

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 8 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: Lucca.Logs/LogConfigurationException.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+
+namespace Lucca.Logs
+{
+    [Serializable]
+    public class LogConfigurationException : Exception
```
Comment (LuccaIntegration) [inline Lucca.Logs/LogConfigurationException.cs]: MAJOR Update this implementation of 'ISerializable' to conform to the recommended serialization pattern. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3925)

File: Lucca.Logs/LuccaDataWrapper.cs [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+            return urlBuilder.ToString();
+        }
+        
+        public static class LogExtractor
+        {
+            public static Func<Exception, IEnumerable<KeyValuePair<string, string>>> CustomKeys { get; set; }
+        }
+
+        [Flags]
+        internal enum UriPart
```
Comment (LuccaIntegration) [inline Lucca.Logs/LuccaDataWrapper.cs]: MINOR Rename this enumeration to match the regular expression: '^([A-Z]{1,3}[a-z0-9]+)*([A-Z]{2})?s$'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2342)

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: Lucca.Logs/LogConfigurationException.cs (hunk 3 above)
Comment (LuccaIntegration) [inline Lucca.Logs/LogConfigurationException.cs]: MAJOR Update this implementation of 'ISerializable' to conform to the recommended serialization pattern. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3925)

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**